### DISCUSSION TITLE
Installing sonarqube via homebrew on Mac presents an error in Wrapper.

### DISCUSSION THREAD

**Discussion by victorhcortesp** [Homebrew/discussions#3022] [Everyday usage]
I have macOS Monterey 12.2.1 version.
MacBook Pro 2019

Perform the installation of sonarqube as indicated in the steps [sonarqube](https://formulae.brew.sh/formula/sonarqube#default).

Additionally I also installed sonar-scanner [sonar-scanner](https://formulae.brew.sh/formula/sonar-scanner#default).

And I get the following error.

```
Running SonarQube...
wrapper  | --> Wrapper Started as Console
wrapper  | Java Service Wrapper Community Edition 64-bit 3.5.48
wrapper  |   Copyright (C) 1999-2021 Tanuki Software, Ltd. All Rights Reserved.
wrapper  |     http://wrapper.tanukisoftware.com
wrapper  |
wrapper  | Launching a JVM...
jvm 1    | WrapperManager: Initializing...
jvm 1    | WrapperManager: ERROR - The version of the Wrapper which launched this JVM is "3.5.48"
jvm 1    | WrapperManager:         while the version of the Wrapper jar file currently in use
jvm 1    | WrapperManager:         is "3.5.49".
jvm 1    | WrapperManager:
jvm 1    | WrapperManager: ERROR - The version of the Wrapper which launched this JVM is
jvm 1    | WrapperManager:         "3.5.48" while the version of the native library
jvm 1    | WrapperManager:         is "3.5.49".
jvm 1    | WrapperManager:
wrapper  | <-- Wrapper Stopped
```

sonarqube  error  256 ~/Library/LaunchAgents/homebrew.mxcl.sonarqube.plist

**Comment by carlocab** [comment]
Does `brew reinstall -s sonarqube` help?

**Reply by MustCodeAl** [reply]
you are a Life saver, im on the m1 mac 11.6.2 big sur

**Comment by cho-m** [comment]
I'm guessing the issue is that Homebrew symlinks the `java-service-wrapper` formula JAR (currently v3.5.49 and will continue updating in future) but copies a specific `wrapper` script (v3.5.48 in last PR https://github.com/Homebrew/homebrew-core/pull/94111) during the bottling process https://github.com/Homebrew/homebrew-core/blob/master/Formula/sonarqube.rb#L29-L32
```rb
    ln_s jsw_libexec/"lib/wrapper.jar", "#{buildpath}/lib/jsw/wrapper-3.5.46.jar"
    ln_s jsw_libexec/"lib/libwrapper.dylib", "#{buildpath}/bin/macosx-universal-64/lib/"
    cp jsw_libexec/"bin/wrapper", "#{buildpath}/bin/macosx-universal-64/"
    cp jsw_libexec/"scripts/App.sh.in", "#{buildpath}/bin/macosx-universal-64/sonar.sh"
```

From error message, there is a hard restriction on version of `wrapper` vs. `wrapper-*.jar` which may break some dependents like `sonarqube` whenever `java-service-wrapper` Formula is updated.

Now, the question is how to fix. Based on comment, it sounds like we cannot symlink the `wrapper` script and so we are locked to version provided at time of build. In this case, we may need to copy both script and JAR to avoid version mismatch. The downside is this is similar to bundling static libraries and bugs/security issues will remain until next update. 

However, unless someone finds a way to bypass symlink restriction, this may be necessary.

**EDIT**: A workaround would be to rebuild dependents. i.e. We could add a testcase in every formula that uses `java-service-wrapper` to try to catch failure during dependent testing and then force rebuild them (whether this is reasonable would depend on cadence of `java-service-wrapper` releases).

**Reply by carlocab** [reply]
> **EDIT**: A workaround would be to rebuild dependents. i.e. We could add a testcase in every formula that uses `java-service-wrapper` to try to catch failure during dependent testing and then force rebuild them (whether this is reasonable would depend on cadence of `java-service-wrapper` releases).

I think this is a good solution.

**Reply by cho-m** [reply]
Another potential solution to explore is whether `wrapper.working.dir=#{Formula["sonarqube"].libexec}/bin/macosx-universal-64` can be used to enable symlinking `wrapper`. (https://wrapper.tanukisoftware.com/doc/english/prop-working-dir.html). May need some extra commands to properly detect all files.

Will require making sure this doesn't break anything else, but I did try a quick hack and got `sonar console` to run:
```diff
-    cp jsw_libexec/"bin/wrapper", "#{buildpath}/bin/macosx-universal-64/"
+    ln_sf jsw_libexec/"bin/wrapper", "#{buildpath}/bin/macosx-universal-64/"
+    open("conf/wrapper.conf", "a") do |conf|
+      conf.write "wrapper.working.dir=#{Formula["sonarqube"].libexec}/bin/macosx-universal-64\n"
+    end
```

Needs to be properly cleaned up and maybe fixed for Linux.
